#162414285 Bookmark articles

Bobsar0 opened the pull request:
#### What does this PR do?
Create, Get and Remove Bookmarks on an Article
#### Description of Task to be completed?
 • Create Bookmark controller and routes
 • Create controller methods (`createOrRemove`, `read`) for performing actions on bookmark resource
 • Modify related model/migration files and associations
 • Write tests
 • Create a helper function to handle check for resource existence
 • Refactor bookmark controller using helper function

#### How should this be manually tested?
 • Clone this branch
 • CD into the repo
 • Run `npm install`
 • Run `npm run migrate && npm run seed`
 • Run `npm run dev`
 • Use postman to check the following endpoints:
```
 • http://localhost/3000/articles/:slug/bookmark (POST OR DELETE);
`Initial request creates bookmark, the next request removes the bookmark..and so on`
 • http://localhost/3000/articles/:slug/bookmarks (GET article's bookmarks);
 • http://localhost/3000/users/:userName/bookmarks (GET user's bookmarks); 
```

#### Background contexts?
Note that you must be an authenticated user to perform endpoint operations.
 • Login with the following credentials to access endpoints
 ``` 
 email: steve@now.com
 password: Blahblah
``` 
 • Use generated token as `bearer token` auth type to access endpoints.

#### What are the relevant pivotal tracker stories?
[Create, Get and Remove Article Bookmarks](https://www.pivotaltracker.com/story/show/162414285)

Commit:
feat(CRUD Bookmark) Perform CRUD operations on Bookmark resource

- Create BookmarkController with methods for performing CRUD operations
- Write tests for the controller
- Modify related model/migration files and associations

finishes #162414285

Commit:
feat(Bookmark Articles): Refactor Bookmark Controller- Create a helper function to handle check for resource existence- Refactor bookmark controller using helper functionfinishes #162414285

jesseinit commented on server/controllers/BookmarkController.js:
```diff
@@ -0,0 +1,127 @@
+      const article = await Util.resourceExists(Article, { slug });
+      if (!article) return response(res, 404, 'failure', 'Article not found');
+
+      Bookmark.findOrCreate({
```
Hi Steve,
Good work on the PR
Any reason why you didn't use sync await on here?

rafmme replied:
@Bobsar0 just a reminder, don't forget to close this conversation when resolved

Bobsar0 replied:
Thanks Jesse. Will resolve

jesseinit commented on server/routes/api/bookmark.js:
```diff
@@ -0,0 +1,13 @@
+const bookmarkRoutes = Router();
+const path = '/articles/:slug/bookmark';
+
+bookmarkRoutes.post(path, AuthMiddleware.checkIfUserIsAuthenticated, BookmarkController.create);
```
Route definition doesn't seem to be in-line with how other routes have been defined and agreed by the team. Any reason for this decision?

Bobsar0 replied:
Yeah. Just following DRY principles. Will change though. Thanks for noticing

tejiri4 commented:
LGTM, yet.

Makwe-O commented:
Apart from the way you defined the routes, This PR looks fine. LGTM Yet

rafmme commented:
LGTM, yet

ebzeal commented:
LGTM yet

Commit:
feat(Bookmark Articles): Add admin auth middleware to getArticleBookmark

- Add admin auth middleware to route while fetching article bookmarks
- Refactor tests
